### PR TITLE
feature: Largest fraction

### LINKED ISSUE #67: feat: add algorithm 'Largest fraction' for Hare & Droop

### LINKED ISSUE #121: refactor: clean up comments in lague-dhont.ts
lague-dhondt.st:20 and utils.ts:140 state that levelling seats has to be == 19. This is made out of date by #120, as the function no longer uses the number of levelling seats, but rely on the number of districts instead.

The original comment was intended to reflect that the algorithm will not behave as originally described when the number of levelling seats != the number of districts.

- [x] "== 19" comment
- [x] const comment

### LINKED ISSUE #146: fix: area factor is always 1.8
When changing area factor, the area factor used for computation does not change.

### PR DESCRIPTION
**Pull request by Log234**
# Description
Implements largest fraction for both Hare and Droop.
Fixes outdated comments in the code.

## Issues closed
closes #121 
closes #67  
closes #146

### CODE REVIEW

**Comment by Log234** [comment]
Currently stuck on the following problem:
1) Distribute the district seats for each district
2) Remove parties that are below the election threshold
3) Distribute the district seats + levelling seats in a national distribution
4) Subtract the number of district seats each party won from the seats they won in the national distribution to get the number of levelling seats each party should win

Problem: 
If the parties that were removed due to the threshold won any district seats, those seats are not subtracted in the final step. Because of this, we may end up with more than the total number of seats.

Status:
I have sent an email to Bernt asking for clarification around the problem.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

File: README.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Lavinia-client
-[![Codacy Badge](https://api.codacy.com/project/badge/Grade/64b563a2c5e349e2a13fcfe3c1bc1008)](https://www.codacy.com/gh/Project-Lavinia/Lavinia-client?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=Project-Lavinia/Lavinia-client&amp;utm_campaign=Badge_Grade) [![Total alerts](https://img.shields.io/lgtm/alerts/g/Project-Lavinia/Lavinia-client.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/Project-Lavinia/Lavinia-client/alerts/)
+
```
Comment (simendag) [inline README.md:2]: Is this space added by you or by our formatting rules? Just curious.

Comment (Log234) [reply README.md:2]: Yes, I did not manually change this, but Codacy seems to agree with the style.

File: src/Layout/Presentation/ConnectedPresentation.tsx [hunk 2]
```diff
@@ -15,6 +15,7 @@ const mapStateToProps = (state: RootState): Partial<PresentationProps> => {
         disproportionalityIndex: state.presentationMenuState.disproportionalityIndex,
         showComparison: state.presentationMenuState.showComparison,
         year,
+        algorithm: state.computationState.algorithm,
```
Comment (simendag) [inline src/Layout/Presentation/ConnectedPresentation.tsx:18]: Why was this change necessary? Because we are now dependent on knowing the algorithm to handle logic further down in more deeply nested components properly?

File: src/Layout/Presentation/DistrictOverview/DistrictOverview.tsx [hunk 3]
```diff
@@ -44,15 +47,7 @@ export class DistrictOverview extends React.Component<DistrictOverviewProps, {}>
...
-                        {" Det mest utsatte sistemandatet (relativt til kvotient) var i "}
-                        {<span className="has-text-warning">{mostVulnerable.district}</span>}
-                        {" og ble vunnet av "}
-                        {mostVulnerable.winner.partyCode}
-                        {". "}
-                        {mostVulnerable.runnerUp.partyCode}
-                        {" ville trengt "}
-                        {mostVulnerable.moreVotesToWin.toFixed(0)}
-                        {" flere stemmer for å vinne det."}
+                        {this.getVulnerableText(mostVulnerable)}
```
Comment (simendag) [inline src/Layout/Presentation/DistrictOverview/DistrictOverview.tsx:50]: Likely a nice, clean change. Though we might consider simply turning it into a component instead.

File: src/Layout/Presentation/ConnectedPresentation.tsx (hunk 2 above)
Comment (Log234) [reply src/Layout/Presentation/ConnectedPresentation.tsx:18]: Yes, we do not want to calculate closest seat when using largest fraction, Celius does not calculate that either.

File: src/Layout/Presentation/DistrictOverview/DistrictOverview.tsx [hunk 4]
```diff
@@ -30,7 +32,8 @@ export class DistrictOverview extends React.Component<DistrictOverviewProps, {}>
                 {data.find((entry) => entry.votesPerSeat === lowestVotingPower)!.name}
             </span>
         );
-        const mostVulnerable = getMostVulnerableSeatByQuotient(data);
+        const calculateVulnerable = isQuotientAlgorithm(this.props.algorithm);
+        const mostVulnerable = calculateVulnerable ? getMostVulnerableSeatByQuotient(data) : undefined;
```
Comment (simendag) [inline src/Layout/Presentation/DistrictOverview/DistrictOverview.tsx:36]: Why does that return "undefined"? That seems suspicious.

File: src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx [hunk 5]
```diff
@@ -1,7 +1,8 @@
 import * as React from "react";
 import ReactTable, { Column } from "react-table";
-import { LevelingSeat, DistrictResult, DistrictQuotients } from "../../../computation";
+import { LevelingSeat, DistrictResult, DistrictQuotients, AlgorithmType } from "../../../computation";
 import { norwegian } from "../../../utilities/rt";
+import { isQuotientAlgorithm } from "../../../computation/logic";
```
Comment (simendag) [inline src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx:5]: I feel like this is indicative of a problem with our architecture. It's strange that logic is computed and embedded here rather than the output simply being displayed.

File: src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx [hunk 6]
```diff
@@ -36,11 +38,12 @@ export class SingleDistrict extends React.Component<SingleDistrictProps, {}> {
...
+        const calculateVulnerable = isQuotientAlgorithm(this.props.algorithm);
         const currentDistrictResult = this.getDistrictResult(this.props.districtSelected);
-        const vulnerableMap = getVotesToVulnerableSeatMap(currentDistrictResult!);
-        const quotientMap = getQuotientsToVulnerableSeatMap(currentDistrictResult!);
-        const vulnerable = getVulnerableSeatByQuotient(currentDistrictResult!);
-        const vulnerableVotes = getVulnerableSeatByVotes(currentDistrictResult!);
+        const vulnerableMap = calculateVulnerable ? getVotesToVulnerableSeatMap(currentDistrictResult!) : undefined;
+        const quotientMap = calculateVulnerable ? getQuotientsToVulnerableSeatMap(currentDistrictResult!) : undefined;
+        const vulnerable = calculateVulnerable ? getVulnerableSeatByQuotient(currentDistrictResult!) : undefined;
+        const vulnerableVotes = calculateVulnerable ? getVulnerableSeatByVotes(currentDistrictResult!) : undefined;
```
Comment (simendag) [inline src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx:46]: Again, weird undefined returns.

File: src/Layout/Presentation/DistrictOverview/DistrictOverview.tsx (hunk 4 above)
Comment (Log234) [reply src/Layout/Presentation/DistrictOverview/DistrictOverview.tsx:36]: We return undefined when the algorithm is not a quotient based algorithm, as the closest seat does not always make sense for largest fraction. For example, all seats may be won in a single round, there is no definition of which seat was last.

File: src/computation/logic/algorithm-utilities.ts [hunk 7]
```diff
@@ -138,12 +167,30 @@ export function getDenominator(algorithm: AlgorithmType, numberOfSeatsAssigned:
             }
         case AlgorithmType.D_HONDT:
             return numberOfSeatsAssigned + 1;
+        case AlgorithmType.LARGEST_FRACTION_HARE:
+        case AlgorithmType.LARGEST_FRACTION_DROOP:
+        case AlgorithmType.LARGEST_FRACTION_HAGENBACH_BISCHOFF:
+            return getElectionNumber(algorithm, totalVotes, totalSeats);
         default:
             console.error(`ERROR! ${algorithm.toString()} does not have an associated denominator function!`);
```
Comment (simendag) [inline src/computation/logic/algorithm-utilities.ts:175]: This is missing the fancy function that checks the type for all cases being checked.

File: src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx (hunk 5 above)
Comment (Log234) [reply src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx:5]: It is a result of us calculating the closest seat on the display side, rather than in the logic.

File: src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx (hunk 6 above)
Comment (Log234) [reply src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx:46]: See my earlier comment for an answer to this.

File: src/computation/logic/algorithm-utilities.ts (hunk 7 above)
Comment (Log234) [reply src/computation/logic/algorithm-utilities.ts:175]: On it.

File: src/computation/logic/distribute-levelling-seats.ts [hunk 8]
```diff
@@ -2,67 +2,42 @@ import { ComputationPayload, PartyResult, DistrictResult, PartyRestQuotients, Re
...
-
-    // Compute the distribution of the total number of seats on the whole country
-    const nationalDistribution = distributeSeats(
-        payload.algorithm,
-        payload.firstDivisor,
-        Number.MIN_SAFE_INTEGER,
-        seatsToDistribute,
-        levelingParties
+    // Filter out parties with less than the threshold
+    let levelingPartyCodes = Object.keys(partyResults).filter(
```
Comment (simendag) [inline src/computation/logic/distribute-levelling-seats.ts:23]: I don't like this mutability.

File: src/computation/logic/distribute-levelling-seats.ts [hunk 9]
```diff
@@ -90,6 +65,164 @@ export function distributeLevelingSeats(
...
+ *
+ * @param levelingPartyCodes The party codes to filter
+ * @param payload The computation payload
+ * @param partyResults The party results for each party
+ */
+function nationalDistributionFilter(
+    levelingPartyCodes: string[],
+    payload: ComputationPayload,
+    partyResults: Dictionary<PartyResult>
+): { nationalDistribution: DistributionResult; levelingPartyCodes: string[] } {
```
Comment (simendag) [inline src/computation/logic/distribute-levelling-seats.ts:84]: Shouldn't this be typed more explicitly?

File: src/computation/logic/distribute-levelling-seats.ts [hunk 10]
```diff
@@ -90,6 +65,164 @@ export function distributeLevelingSeats(
...
+    let localLevelingPartyCodes = [...levelingPartyCodes];
+    let nationalDistribution: DistributionResult;
+
+    let unfilteredPartiesLength;
+    let filteredPartiesLength = localLevelingPartyCodes.length;
```
Comment (simendag) [inline src/computation/logic/distribute-levelling-seats.ts:89]: Mutability.

File: src/computation/logic/distribute-levelling-seats.ts [hunk 11]
```diff
@@ -90,6 +65,164 @@ export function distributeLevelingSeats(
...
+                partyId: -1,
+                resultId: -1,
+                countyName: "",
+                partyCode,
+                partyName: "",
+                votes: partyResults[partyCode].votes,
+                percentage: -1,
+            };
+            levelingParties.push(party);
+        }
```
Comment (simendag) [inline src/computation/logic/distribute-levelling-seats.ts:111]: For a later time, I think this could probably be achieved by a few rewrites and a map of some kind (assuming it doesn't harshly affect performance).

File: src/computation/logic/distribute-levelling-seats.ts [hunk 12]
```diff
@@ -90,6 +65,164 @@ export function distributeLevelingSeats(
...
+    levelingPartyCodes: string[],
+    payload: ComputationPayload,
+    partyResults: Dictionary<PartyResult>
+): { nationalDistribution: DistributionResult; levelingPartyCodes: string[] } {
+    let localLevelingPartyCodes = [...levelingPartyCodes];
+    let nationalDistribution: DistributionResult;
+
+    let unfilteredPartiesLength;
+    let filteredPartiesLength = localLevelingPartyCodes.length;
+    do {
```
Comment (simendag) [inline src/computation/logic/distribute-levelling-seats.ts:90]: I am very skeptical to the usage of do-whiles, they are generally just added complexity.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

File: src/computation/logic/lague-dhondt.ts [hunk 13]
```diff
@@ -139,7 +136,7 @@ export function lagueDhont(payload: ComputationPayload): LagueDhontResult {
     const districtResultArray = dictionaryToArray(districtResults);
     const partyResultArray = dictionaryToArray(partyResults);
 
-    const useAdjustedQuotients = payload.parameters.electionYear >= 2005;
+    const useAdjustedQuotients = payload.parameters.electionYear >= 2005 && isQuotientAlgorithm(payload.algorithm);
```
Comment (simendag) [inline src/computation/logic/lague-dhondt.ts:139]: 2005 is a magic number.

File: src/computation/logic/quotient-dictionary.ts [hunk 14]
```diff
@@ -34,7 +35,7 @@ export class QuotientDictionary extends SortedReverseDict {
         // If more than 1 districts share the lead
         if (winners.length > 1) {
             // Select a winner
-            winner = this.tieBreaker(winners, baseValue);
+            winner = tieBreaker(winners, baseValue);
```
Comment (simendag) [inline src/computation/logic/quotient-dictionary.ts:38]: A function name should arguably be a verb rather than a noun. "breakTie(winners, baseValue)" for instance. `baseValue` is a little cryptic.

File: src/computation/logic/sorted-reverse-dict.ts [hunk 15]
```diff
@@ -16,7 +16,7 @@ export class SortedReverseDict {
         const start = this.reverseDict[0].value;
         let num = 0;
 
-        while (this.reverseDict[num].value === start) {
+        while (num < this.reverseDict.length && this.reverseDict[num].value === start) {
```
Comment (simendag) [inline src/computation/logic/sorted-reverse-dict.ts:19]: The conditional is a little unclear and complex here. Could be clarified by variables above and a more natural language conditional put into the while.

File: src/computation/logic/utils.ts [hunk 16]
```diff
@@ -185,7 +186,7 @@ export function distributeLevelingSeatsOnDistricts(
         if (levelingSeats.length === 0) {
             finishedDistricts = [];
             levelingSeats = generateLevelingSeatArray(
-                payload.algorithm,
+                AlgorithmType.SAINTE_LAGUE,
```
Comment (simendag) [inline src/computation/logic/utils.ts:189]: Are leveling seat arrays always SAINTE_LAGUE?

File: src/computation/logic/distribute-levelling-seats.ts (hunk 8 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts:23]: Fixed.

File: src/computation/logic/distribute-levelling-seats.ts (hunk 9 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts:84]: Fixed.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

File: src/computation/logic/distribute-levelling-seats.ts (hunk 10 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts]: Fixed by making the function recursive.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

File: src/computation/logic/distribute-levelling-seats.ts (hunk 11 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts]: I had a look at it. The issue is that Dictionary is not iterable, so we cannot easily use map or reduce without first converting it to an array.

File: src/computation/logic/distribute-levelling-seats.ts (hunk 12 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts]: Made the function recursive instead.

File: src/computation/logic/lague-dhondt.ts (hunk 13 above)
Comment (Log234) [reply src/computation/logic/lague-dhondt.ts:139]: Do you have any suggestions for how to fix this? Hard to find a nice descriptive name for a variable or function.

File: src/computation/logic/quotient-dictionary.ts (hunk 14 above)
Comment (Log234) [reply src/computation/logic/quotient-dictionary.ts:38]: Fixed.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

File: src/computation/logic/sorted-reverse-dict.ts (hunk 15 above)
Comment (Log234) [reply src/computation/logic/sorted-reverse-dict.ts]: Renamed the variables, should make it a bit more intuitive.

File: src/computation/logic/utils.ts (hunk 16 above)
Comment (Log234) [reply src/computation/logic/utils.ts:189]: Bernt said we could always use Sainte Laguës for distribution leveling seats on each district.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

File: src/computation/logic/utils.ts (hunk 16 above)
Comment (Log234) [reply src/computation/logic/utils.ts:189]: The whole function should really be rewritten.

File: src/computation/logic/algorithm-utilities.ts (hunk 7 above)
Comment (simendag) [reply src/computation/logic/algorithm-utilities.ts:176]: Did this get resolved? I don't see it in the default case.

File: src/computation/logic/distribute-levelling-seats.ts [hunk 17]
```diff
@@ -2,15 +2,94 @@ import { ComputationPayload, PartyResult, DistrictResult, PartyRestQuotients, Re
...
+            partyResults[partyCode].levelingSeats += levelingSeats;
+            partyResults[partyCode].totalSeats += levelingSeats;
+        }
+    }
+
+    const wonLevelingPartyCodes = levelingPartyCodes.filter((partyCode) => partyResults[partyCode].levelingSeats > 0);
+
+    let partyRestQuotients: Dictionary<PartyRestQuotients> = {};
+
+    if (payload.election.year < 2005) {
```
Comment (simendag) [inline src/computation/logic/distribute-levelling-seats.ts:51]: Magic number still. This should probably be a const that we name and export given how often it's used and in how many places.

File: src/computation/logic/distribute-levelling-seats.ts [hunk 18]
```diff
@@ -21,75 +100,124 @@ export function distributeLevelingSeats(
...
+        totalVotes,
         levelingParties
     );
 
-    // Filter out parties who did not gain any seats from the national distribution
-    levelingParties = levelingParties.filter(
-        (party) => nationalDistribution.seatsWon[party.partyCode] >= partyResults[party.partyCode].districtSeats
+    // Filter out parties that did not gain any seats in the new distribution
+    const filteredLevelingPartyCodes = levelingPartyCodes.filter(
+        (p) => nationalDistribution.seatsWon[p] > partyResults[p].districtSeats
```
Comment (simendag) [inline src/computation/logic/distribute-levelling-seats.ts:118]: The logic here changed, is it intentional? `>=` vs `>`

File: src/computation/logic/distribute-levelling-seats.ts [hunk 19]
```diff
@@ -21,75 +100,124 @@ export function distributeLevelingSeats(
...
     );
 
-    let levelingPartyCodes = levelingParties.map((party) => party.partyCode);
+    if (filteredLevelingPartyCodes.length !== levelingPartyCodes.length) {
```
Comment (simendag) [inline src/computation/logic/distribute-levelling-seats.ts:121]: The conditional here isn't very intuitive to me -- what does it actually represent?

File: src/computation/logic/distribute-levelling-seats.ts (hunk 18 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts:118]: Yes, only parties that **gain** seats in the national distribution should be in the draw for levelling seats.

File: src/computation/logic/distribute-levelling-seats.ts [hunk 20]
```diff
@@ -21,75 +100,124 @@ export function distributeLevelingSeats(
...
+    const levelingSeatDistribution: Dictionary<number> = {};
+    levelingPartyCodes.forEach((partyCode) => {
+        levelingSeatDistribution[partyCode] =
+            nationalDistribution.seatsWon[partyCode] - partyResults[partyCode].districtSeats;
+    });
 
-    const levelingSeatDistribution = dictionaryToArray(partyRestQuotients);
-    return levelingSeatDistribution;
+    return {
+        seatResults: [],
```
Comment (simendag) [inline src/computation/logic/distribute-levelling-seats.ts:220]: Why is an empty array being returned here?

File: src/computation/logic/distribute-levelling-seats.ts (hunk 17 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts:51]: I agree, earlier I asked you for a suggestion for a name. It is quite hard to find something that is intuitive and clear, rather than just more confusing.

File: src/computation/logic/utils.ts (hunk 16 above)
Comment (simendag) [reply src/computation/logic/utils.ts:189]: One wonderful day when there is time!

File: src/computation/logic/utils.ts [hunk 21]
```diff
@@ -281,3 +282,24 @@ export function distributeLevelingSeatsOnDistrictsPre2005(
...
     return partyRestQuotients;
 }
+
+/**
+ * Breaks ties in the distribution of items on names
+ *
+ * @param winners The list of multiple winners from the distribution stage
+ * @param baseValue The dictionary from winners to their respective numerators
+ */
+export function breakTies(winners: KeyValuePair[], baseValue: Dictionary<number>): KeyValuePair {
```
Comment (simendag) [inline src/computation/logic/utils.ts:292]: What is this KeyValuePair? That seems suspicious considering there are both Maps and a "custom" Dictionary in our code base already.

File: src/computation/logic/distribute-levelling-seats.ts (hunk 19 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts:121]: If we found a party that should be filtered out, the national distribution should be repeated. Here is an example from Stortinget which uses this behaviour: [https://www.stortinget.no/no/Saker-og-publikasjoner/Publikasjoner/Innstillinger/Stortinget/2005-2006/inns-200506-001/7/](https://www.stortinget.no/no/Saker-og-publikasjoner/Publikasjoner/Innstillinger/Stortinget/2005-2006/inns-200506-001/7/)

File: src/computation/logic/distribute-levelling-seats.ts (hunk 20 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts:220]: Because levelling seats are not distributed in rounds like Sainte Laguës or d'Hondt. So there is not necessarily any definition of who won a round, as all seats may be distributed in a single round.

File: src/computation/logic/distribute-levelling-seats.ts (hunk 19 above)
Comment (simendag) [reply src/computation/logic/distribute-levelling-seats.ts:121]: const existsUnfilteredParty = ...conditional... perhaps?

File: src/computation/logic/utils.ts (hunk 21 above)
Comment (Log234) [reply src/computation/logic/utils.ts:292]: Yes, I noticed it as well. It is tied to the SortedReverseDict, so I thought it would be better to make another PR to simplify the functions of it.

File: src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx (hunk 5 above)
Comment (simendag) [reply src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx:5]: Why do we do that? Do we have to, or is it potentially fixable in the future?

File: src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx (hunk 6 above)
Comment (simendag) [reply src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx:46]: The function itself should probably return undefined, and we could encapsulate all these changes in an if block.

File: src/computation/logic/distribute-levelling-seats.ts (hunk 19 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts:121]: Fixed.

File: src/computation/logic/algorithm-utilities.ts (hunk 7 above)
Comment (Log234) [reply src/computation/logic/algorithm-utilities.ts:176]: Yes, but the fix did not touch the lines 170-173, it happened in 174+.

File: src/computation/logic/utils.ts (hunk 21 above)
Comment (simendag) [reply src/computation/logic/utils.ts:292]: Maybe make an issue out of it?

File: src/computation/logic/distribute-levelling-seats.ts (hunk 19 above)
Comment (simendag) [reply src/computation/logic/distribute-levelling-seats.ts:121]: Did you push the change you made? I can't see it.

File: src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx (hunk 5 above)
Comment (Log234) [reply src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx:5]: Good question, I did not write that code. I believe you did.

File: src/computation/logic/distribute-levelling-seats.ts (hunk 17 above)
Comment (simendag) [reply src/computation/logic/distribute-levelling-seats.ts:51]: Well, if we stay in the scope of this bit of code without thinking about exports, couldn't we do `const preOneLevellingSeatPerDistrict = payload.election.year < 2005`? It's a little bulky but does make the intention quite clear.

File: src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx (hunk 6 above)
Comment (Log234) [reply src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx:46]: What do you mean by encapsulating all the changes in an if block?

File: src/Layout/Presentation/DistrictOverview/DistrictOverview.tsx (hunk 3 above)
Comment (Log234) [reply src/Layout/Presentation/DistrictOverview/DistrictOverview.tsx]: Fixed.

File: src/computation/logic/distribute-levelling-seats.ts (hunk 19 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts]: Pushed two changes now, see if it updated.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

File: src/computation/logic/distribute-levelling-seats.ts (hunk 17 above)
Comment (Log234) [reply src/computation/logic/distribute-levelling-seats.ts:51]: Sure, I was thinking more on a repository-wide parameter, since it is a conditional that appears in a lot of different places. But finding a name per scope is not quite as challenging.

File: src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx (hunk 6 above)
Comment (simendag) [reply src/Layout/Presentation/SingleDistrict/SingleDistrict.tsx:46]: Given that there are 4 lines using the same conditional (which should probably be renamed to shouldCalculateVulnerable) we might consider having two returns (one for each case) and rewriting a bit of the code to use components, but I don't think this is important right now. We can create clean up issues when/if we have time for something as small as this.

File: src/computation/logic/lague-dhondt.ts (hunk 13 above)
Comment (Log234) [reply src/computation/logic/lague-dhondt.ts]: Cleaned up quite a few conditionals.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           

Complexity increasing per file
==============================
- src/utilities/conditionals.ts  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

File: src/computation/logic/utils.ts (hunk 21 above)
Comment (Log234) [reply src/computation/logic/utils.ts:292]: Created issue #144

File: src/utilities/conditionals.ts [hunk 22]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Determines wether or not district seats should be distributed for a given year.
+ *
+ * @param year The year to check
+ */
+export function distributeDistrictSeats(year: number): boolean {
```
Comment (simendag) [inline src/utilities/conditionals.ts:6]: Maybe better to have `shouldDistributeDistrictSeats` -- right now the function name indicates that it _does_ distribute district seats. I find it's nice to have the first segmented word immediately make you think of a boolean.

File: src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx [hunk 23]
```diff
@@ -65,10 +65,11 @@ export class RemainderQuotients extends React.Component<RemainderQuotientsProps>
...
                     if (row.value !== undefined) {
                         let quotient = row.value.quotient;
-                        if (this.props.year < 2005 && isQuotientAlgorithm(this.props.algorithm)) {
-                            quotient = quotient / 10000;
-                        } else if (this.props.year >= 2005 && isQuotientAlgorithm(this.props.algorithm)) {
+                        const useAdjustedQuotient = this.props.year >= 2005;
+                        if (useAdjustedQuotient && isQuotientAlgorithm(this.props.algorithm)) {
                             quotient = quotient * 10000;
+                        } else if (isQuotientAlgorithm(this.props.algorithm)) {
+                            quotient = quotient / 10000;
```
Comment (simendag) [inline src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx:72]: Remind me what the significance of `10000` is here?

Comment (Log234) [reply src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx:72]: It is related to whether or not adjusted quotients were used. But it seems a bit strange, this division/multiplication makes quotients for some years in the range of 1000-10 000, and others 0.9-0.01.

Comment (simendag) [reply src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx:72]: Isn't there a name for such a constant?

File: src/utilities/conditionals.ts (hunk 22 above)
Comment (Log234) [reply src/utilities/conditionals.ts]: That is indeed a better name. Fixed.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           

Complexity increasing per file
==============================
- src/utilities/conditionals.ts  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

File: src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx (hunk 23 above)
Comment (Log234) [reply src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx:72]: A scalar perhaps?

Comment (simendag) [reply src/Layout/Presentation/RemainderQuotients/RemainderQuotients.tsx:72]: I meant what is its role in the formula? Anyway, it will likely be picked up when we set up goals for refactoring at some point, and doesn't have to be dealt with for this release.

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           

Complexity increasing per file
==============================
- src/utilities/conditionals.ts  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           

Complexity increasing per file
==============================
- src/utilities/conditionals.ts  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)

**Comment by Log234** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
           

Complexity increasing per file
==============================
- src/utilities/conditionals.ts  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Project-Lavinia/Lavinia-client/pullRequest?prid=5453100&bid=17878092)
